### PR TITLE
Fix duplicates and whitesource

### PR DESCRIPTION
**Pull request by MarianG**
We had an issue where we encountered circular dependencies for duplicates in findings. We assume it occured because of multi-threading and the asynchronous duplicate-detection. We had issues where Finding A had duplicates B, C, and D. B had the duplicate C, C had the duplicate D and D had B. Then the clean-up job that should delete some duplicates produced a recursion error, because it tries to delete duplicates of a finding before deleting the finding. 
I understand that in theory a duplicate should not have a duplicate on its own, but it happened.
We updated the *set_duplicate* method and included a way to identify and repair loops. 
We also found the duplicate_list field is unnecessary, since we have the information in the original_finding field (backwards from duplicate_finding Foreign Key)

The other change is for the whitesource parser. We fixed an issue where the parser fails when the CVE is not set. Additionally the updated title resulted in problems for the duplicate detection of old findings, since now the lib_name is used, which changed the title. We included a method to recreate all old titles and recalculate the hash to make duplication work again.

On behalf of DB Systel GmbH

**Note: DefectDojo is now on Python3 and Django 2.2.1 Please submit your pull requests to the 'dev' branch as the 'legacy-python2.7' branch is only for bug fixes. Any new features submitted to the legacy branch will be ignored and closed.**

When submitting a pull request, please make sure you have completed the following checklist:

- [x] Your code is flake8 compliant 
- [x] Your code is python 3.5 compliant
- [ ] If this is a new feature and not a bug fix, you've included the proper documentation in the ReadTheDocs documentation folder. https://github.com/DefectDojo/Documentation/tree/master/docs or provide feature documentation in the PR.
- [x] Model changes must include the necessary migrations in the dojo/dd_migrations folder.
- [x] Add applicable tests to the unit tests.

### CODE REVIEW

File: dojo/tasks.py [hunk 1]
```diff
@@ -291,20 +311,20 @@ def async_update_findings_from_source_issues(*args, **kwargs):
...
-            duplicate_list = finding.duplicate_list \
+            duplicate_list = finding.original_finding \
                     .filter(duplicate=True).order_by('date')
             dupe_count = len(duplicate_list) - dupe_max
             for finding in duplicate_list:
-                deduplicationLogger.debug('deleting finding {}:{} ({}))'.format(finding.id, finding.title, finding.hash_code))
+                logger.info("Debug-Dupe: Delete Finding with id: %d", finding.id)
                 finding.delete()
                 dupe_count = dupe_count - 1
                 if dupe_count == 0:
```
Comment (valentijnscholten) [inline dojo/tasks.py]: I think there is an unintended change from `deduplicationLogger` to `logger`?

**Comment by valentijnscholten** [comment]
Would be good if @ptrovatelli takes a look (the king of dedupe).

File: dojo/tasks.py (hunk 1 above)
Comment (MarianG) [reply dojo/tasks.py]: ah yes. I think this can be changed back

Comment (MarianG) [reply dojo/tasks.py]: should i revert this to deduplicationLogger?

**Comment by madchap** [comment]
I don't know enough of the internal dedup, but I welcome the dedup tests for sure.

The whitesource change is minor and is fine. The renaming piece looks OK as well.

**Comment by MarianG** [comment]
I am also not sure why the smoke test fails. If I run this locally it looks good.

**Comment by madchap** [comment]
So, should this PR now be closed in favor of #2009, #2010 and #2011 ?

**Comment by MarianG** [comment]
yes I think we can close. I just wanted to wait until the tests, for the other PR's #2009 #2010 #2011 are successful

**Pull request closed without merging**